### DISCUSSION TITLE
Setting a global variable by an invoked state machine

### DISCUSSION THREAD

**Discussion by larry104** [alexzhornyak/SCXML-tutorial#24] [Q&A]
I have currently something like this where I track key presses
![image](https://user-images.githubusercontent.com/2013720/231276881-f9e018e6-3f65-46cf-bc05-ac1e0673617e.png)

I thought I replace it with a state machine which I invoke to make it easier since I needs about 20 of these.
![image](https://user-images.githubusercontent.com/2013720/231277146-40031a9b-72a3-4b4e-9b06-c41599dac0a4.png)

And the corresponding invoked state_machine looks like this:
![image](https://user-images.githubusercontent.com/2013720/231277262-73503a0e-c9e5-40bf-b92f-4f0e93dea35b.png)

The variable VarKeyState keeps track the state of all keys e.g. 0000000010101110 (every bit is one of the keys) 
Unfortunately, the invoked state machine does not know the variable VarKeyState 

Any idea how to get access to a global variable from inside an invoked state machine?

**Comment by alexzhornyak** [comment]
Assignment of global variable should be implement in the parent state machine by event that was send from invoked machine

https://user-images.githubusercontent.com/18611095/231379788-92d6cdf6-8e67-4021-9a31-41c9f2e247ab.mp4

* **Parent Machine**
![image](https://user-images.githubusercontent.com/18611095/231379884-fdfd0f6e-b560-4b41-a58c-8ba4e174c764.png)

* **Child Machine**
![image](https://user-images.githubusercontent.com/18611095/231380010-294a5824-048e-47ef-97e4-58ccbb624bb6.png)

**P.S. And for such very simple case I am not sure that you need invoked machines**

https://user-images.githubusercontent.com/18611095/231381486-348eb031-4a9d-490b-b8f4-33e2c0483f7e.mp4

![image](https://user-images.githubusercontent.com/18611095/231381563-6de098ef-61fc-4273-85fc-203133b6b1d2.png)

**Comment by alexzhornyak** [comment]
[InvokeGlobalVar.zip](https://github.com/alexzhornyak/SCXML-tutorial/files/11208318/InvokeGlobalVar.zip)

**Comment by larry104** [comment]
Wow, thank to so much for your support. That makes a lot of sense to send an event from the child to the parent. Here a few questions though:
1)  i_KEY, t_KEY_STATE, starting with i_, t_  is that just a naming convention?

2) I need to store the status of the keys as bits cleared or set bits in an integer and not in an array. The reason is that it's very easy to check if a certain group of keys are pressed or not. Unfortunately, it seems using the left shift operator will trigger an error 

`<script>key = Number(_event.name.replace('do.key.pressed.KB.', ''))-1
bit = (1<<key)
if(_event.data==0) {
// clear bit
}else{
// set bit
}
</script>
<log expr="t_KEY_KB_STATE.toString(2)" label="Key State"/>
`

`ERROR> Apr 12 12:01:27 Transition[StateMachineConnection42__40E5FC8E9CDB3193] XML Check> Name should began with the letter or symbol "_"`

Is there a trick around that error?

**Reply by alexzhornyak** [reply]
1. Just naming convention
2.  In your case is pretty clear explained that 'Name should began with the letter'. It seems like you directly edit XML not escaping the characters '<<' etc. Use the corresponding GUI editors

![Screenshot 2023-04-12 at 22 49 47](https://user-images.githubusercontent.com/18611095/231569226-5a6f4beb-d061-419e-9e1d-25ec46ab03b4.png)

3. Here I changed code to handle with bits

![image](https://user-images.githubusercontent.com/18611095/231568641-f304472d-5bac-4683-9025-121a45eb1ac7.png)

```xml
<scxml datamodel="ecmascript" name="ScxmlParent" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<datamodel>
		<data expr="0" id="i_BIT_STATE"/>
	</datamodel>
	<parallel id="wrapper">
		<transition event="do.key.click.*">
			<script>var bit = Number(_event.name.replace('do.key.click.', '')) - 1
i_BIT_STATE = _event.data == 1 ? (i_BIT_STATE | 1&lt;&lt;bit) : (i_BIT_STATE &amp; ~(1&lt;&lt;bit))
			</script>
			<log expr="i_BIT_STATE.toString(2)" label="Key State"/>
		</transition>
		<state id="child1">
			<invoke autoforward="true" id="ID_1" src="machine_child.scxml">
				<param expr="1" name="i_KEY"/>
			</invoke>
		</state>
		<state id="child2">
			<invoke autoforward="true" id="ID_2" src="machine_child.scxml">
				<param expr="2" name="i_KEY"/>
			</invoke>
		</state>
	</parallel>
</scxml>
```

**Comment by larry104** [comment]
Cool! Yes, that works. I didn't see there is a special XML Test field as well. The ScxmlEditor is a very powerful tool! It needs a bit of learning but then it's awesome.

Thanks!

**Comment by larry104** [comment]
When changing to invoked statemachines I came across another question/problem ... Are states from the parent (or full hierarchy) known inside the invoked statemachine?  It seems the statemachine in my case is stuck in StateShape10 so the In() doesn't seem to work. 

![image](https://user-images.githubusercontent.com/2013720/231906671-ce276892-4a41-4816-a8a1-f7f0ed9a521e.png)

**Reply by alexzhornyak** [reply]
Invoked machine is isolated session. It does not know about other state machines, about data in other machine scripts.
So, if you should to inform child machines by events

https://user-images.githubusercontent.com/18611095/232038975-2ecb30cf-b320-4a54-8b95-8325521fb726.mp4

[InvokeInState.zip](https://github.com/alexzhornyak/SCXML-tutorial/files/11232593/InvokeInState.zip)

**Comment by larry104** [comment]
I see. So, invokes are like function calls and they only know their local scope. 

Just a random thought ...

When I was writing xml scxml in qt by hand before knowing the ScxmlEditor existed I noticed that invoked statemachines didn'twork properly in qt because qscxmlc gives you an error if you include the same statemachine more than once (at least in qt <=5.7). The way I solved this was by pasting the scxml block I needed to duplicate into a script and then replicating it n number of times of course changing id's and names of events so everything stays unique. The generated scxml block I pasted back into the original scxml. Painful, but it worked. 

I could envision a different way of doing this:

Take for instance your "Unoptimized statechart without invoke" example you have [here ](https://github.com/alexzhornyak/SCXML-tutorial/blob/master/Doc/invoke.md). If you make a change to one of the 4 blocks you have to change the other 3 the same way. So, in situations where you have to be in the scope of the entire statechart and can't use an invoke you could have a new feature which treats one of the 4 blocks as master and the other 3 as a clones. State names like On_1, On_2, On_3,... would need to be defined as On_$ and $ will be replaced when ScxmlEditor dynamically expands the code during runtime into 4 unique blocks. The master could be kept in a separate scxml file and instead of an "invoke" you have to use a "clone" element. Does that kind of make sense? Do you think that would be a useful feature others would be interested in? The other way to see that is having a special "invoke" which is instead of instantiating it flattens out the scxml and so it stays within the scope of the overall statechart.

**Comment by alexzhornyak** [comment]
1. Regarding Qt invoked machines it seems like you are talking about this bug [https://bugreports.qt.io/browse/QTBUG-90544](https://bugreports.qt.io/browse/QTBUG-90544)
2. Invoked machines were designed to be able for using in distributed systems. For example to communicate via HTTP and be launched on another machine.
If you need to speed up your chart and make it with less events, I can offer to declare on the C-side functions that will be able to get-set global variables, to track entered states, etc.

Here is an example with setting global variables with extended **GetGlobal** and **SetGlobal** functions

https://user-images.githubusercontent.com/18611095/232486789-e03ac327-e2c8-44bc-9b03-73ccd010e324.mp4

```xml
<scxml datamodel="luavia" name="ScxmlChild" version="1.0" xmlns="http://www.w3.org/2005/07/scxml">
	<state id="child">
		<onentry>
			<script>i_global = GetGlobal('GlobalVar')
i_global = i_global + 1
SetGlobal('GlobalVar', i_global)
			</script>
		</onentry>
		<transition target="FinalShape1"/>
	</state>
	<final id="FinalShape1"/>
</scxml>
```

[InvokeGetSetGlobal.zip](https://github.com/alexzhornyak/SCXML-tutorial/files/11250211/InvokeGetSetGlobal.zip)

**Reply by larry104** [reply]
1) No, it's a different bug. When you invoke the same statemachine more than once it does not compile anymore. qt seems to define the class name of the child statemachine multiple times. Do you know if they fixed the problem in 6.4? I am very reluctant to upgrade from 5.7 to 6.4 since I have a working crosscompiling flow from Ubuntu to a RaspberryPi ... and it was a pain to set that up. But I was wondering if it makes sense to use uscxml. Do you have an example for Linux somewhere how to interact with qt sending signals from uscxml events and forwarding signals to uscxml? I saw there are examples for Windows using the UscxmlCLib  but I could not find any example using uscxml directly. 
2) Does that work for ecmascript as well?

**Comment by alexzhornyak** [comment]
I have a couple of projects on Raspberry and my workflow is the next:
- design project under Windows using git version control
- on Raspberry I update git project and build finally directly there
- I'm using this Qt-Raspberry 5.15 builds [https://github.com/tranter/raspberry-pi-qt-builds](https://github.com/tranter/raspberry-pi-qt-builds)

Unfortunately I don't have ready-to-use examples in combination of Qt and uscxml. Only through my UscxmlCLib for educational purposes.
There are some issues with uscxml: official uscxml has critical memory leaks, so I'm using my own fork. 

_Does that work for ecmascript as well?_
Yes, watch this example [https://github.com/alexzhornyak/SCXML-tutorial/tree/master/Examples/Qt/SkodaBoleroInfotainment](https://github.com/alexzhornyak/SCXML-tutorial/tree/master/Examples/Qt/SkodaBoleroInfotainment)

![image](https://user-images.githubusercontent.com/18611095/232699617-1f14a3c5-afc1-4812-92d1-7a7250cfa5e8.png)

![image](https://user-images.githubusercontent.com/18611095/232700606-25d25c0a-5da3-4199-8dc5-24b80f91efac.png)

This class is passed to state machine and its methods are used there

![image](https://user-images.githubusercontent.com/18611095/232700360-a431b30d-d9a0-4a5c-809c-af0adbeadd48.png)

**Reply by larry104** [reply]
Thanks a lot. Let me bring up a VM with Ubuntu 22.04 and test if qt 6.5 still has the problems with nested statemachines. If it does I need to switch to uscxml. I really like your ScxmlEditor as it allows me to pretty much design and debug the entire statemachine upfront.
